File: src/branchnexus/retry.py
"""Retry/backoff helpers for recoverable operations."""

from __future__ import annotations

import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import TypeVar

T = TypeVar("T")


class RecoverableError(Exception):
    """Transient failure that can be retried."""


class FatalError(Exception):
    """Non-recoverable failure that should stop immediately."""


@dataclass(frozen=True)
class RetryPolicy:
    max_attempts: int = 3
    initial_backoff_seconds: float = 0.5
    multiplier: float = 2.0


def run_with_retry(
    operation: Callable[[], T],
    *,
    policy: RetryPolicy,
    sleep: Callable[[float], None] = time.sleep,
) -> T:
    attempt = 0
    backoff = policy.initial_backoff_seconds
    last_error: Exception | None = None

    while attempt < policy.max_attempts:
        attempt += 1
        try:
            return operation()
        except FatalError:
            raise
        except RecoverableError as exc:
            last_error = exc
            if attempt >= policy.max_attempts:
                break
            sleep(backoff)
            backoff *= policy.multiplier

    if last_error is not None:
        raise last_error
    raise RuntimeError("Retry policy exhausted without executing operation.")

File: ts-src/utils/retry.ts
export class RecoverableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'RecoverableError';
  }
}

export class FatalError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FatalError';
  }
}

export interface RetryPolicy {
  maxAttempts: number;
  initialBackoffMs: number;
  multiplier: number;
}

export const DEFAULT_RETRY_POLICY: RetryPolicy = {
  maxAttempts: 3,
  initialBackoffMs: 500,
  multiplier: 2,
};

export async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function runWithRetry<T>(
  operation: () => Promise<T>,
  policy: RetryPolicy = DEFAULT_RETRY_POLICY
): Promise<T> {
  let attempt = 0;
  let backoff = policy.initialBackoffMs;
  let lastError: Error | null = null;

  while (attempt < policy.maxAttempts) {
    attempt++;
    try {
      return await operation();
    } catch (error) {
      if (error instanceof FatalError) {
        throw error;
      }
      if (error instanceof RecoverableError) {
        lastError = error;
        if (attempt >= policy.maxAttempts) {
          break;
        }
        await sleep(backoff);
        backoff *= policy.multiplier;
      } else {
        throw error;
      }
    }
  }

  if (lastError) {
    throw lastError;
  }
  throw new Error('Retry policy exhausted without executing operation.');
}
